### PR TITLE
Add ENS resolver records + ENS labels

### PR DESCRIPTION
**Pull request by 0xRobin**
1) Adds new tables for the resolver addresses of ENS names.
For now only baisc (.eth, no subdomains) ENS names are supported.
2) Adds the ENS name models to the label model

ENS main issue: https://github.com/duneanalytics/spellbook/issues/1666

---
*For Dune Engine V2*
I've checked that:
General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)

Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by 0xRobin** [comment]
All relevant tests are passing. :heavy_check_mark: 
This is ready for review @soispoke.
This will also solve https://github.com/duneanalytics/spellbook/issues/1651 @bh2smith

**Review by bh2smith — COMMENTED** [review]
Looks pretty great, just a few comments and questions.

File: models/ens/ens_node_names.sql [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
+    {% endif %}
+)
+
+-- here's the sketchy matching
+, matching as (
+    select *
+    from (
+        select *
+        ,row_number() over (partition by node order by block_time desc, evt_index desc) as ordering2
```
[the remark this answers is not in the document]
Comment (0xRobin) [reply models/ens/ens_node_names.sql:51]: I though about this too but went with only latest state.
- The full history can be compiled quite easy by querying the NewAddr events on the specific Node.
- Having only the latest state in the table improves query speed for handling the labels.

It's deff a trade-off.

File: models/ens/ens_node_names.sql [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    node
+    ,concat(label_name,'.eth') as name
+    ,label_name
+    ,label_hash
+    ,address as initial_address
+    ,tx_hash
+    ,block_number
+    ,block_time
+    ,evt_index
+from matching
```
[the remark this answers is not in the document]
Comment (0xRobin) [reply models/ens/ens_node_names.sql:79]: We don't need to.
From the ENS [docs](https://docs.ens.domains/frequently-asked-questions):
> The released name continues to resolve your ETH address until the new owner overwrites it.

**Comment by soispoke** [comment]
@0xRobin After removing duplicates from other labels sources, lgtm! Thanks again, this is awesome

**Review by soispoke — APPROVED** [review]

**Comment by bh2smith** [comment]
Just wanted to point out that some old ens names are still not appearing here:

Examples:
- [vitalik.eth](https://app.ens.domains/name/vitalik.eth/details)
- [bensmith.eth](https://app.ens.domains/name/bensmith.eth/details) 
- [bahman.eth](https://app.ens.domains/name/bahman.eth/details)


Essentially anything that was registered before ens redeployed their contracts to contain enough event data to actually link all the relevant details.

There was this table in V1 called "old ens preimages" that I believe was used for this purpose:

https://github.com/duneanalytics/spellbook/blob/main/deprecated-dune-v1-abstractions/ethereum/ethereumnameservice/old_preimages.sql


You will find that all the ens names in this list are not appearing in the the table created here and the only way to include that in Dune is to hard code it. 


Here is a query where you can see the examples above are missing.

I went and tried to check a few more addresses from the list linked above, but discovered it is not true that they are all missing from this table. However, it is true that each of the three examples above is contained in this list.

**Comment by 0xRobin** [comment]
That's correct, current impl only includes addresses from ens.view_registrations

**Comment by bh2smith** [comment]
Doesn't look like it. Maybe we could add it to the todo list on this epic endeavour.

**Comment by soispoke** [comment]
Noted, thanks @bh2smith! Yeah definitely adding it to the to do list 👍 Will look into it

**Pull request merged**